app/src/main/java/app/revanced/manager/ui/screen/PatcherScreen.kt

```diff
@@ -0,0 +1,433 @@
+                }
+
+                StepIcon(
+                    state = state,
+                    size = 24.dp
+                )
+
+                Text(
+                    stringResource(category.displayName),
+                    style = MaterialTheme.typography.titleMedium
+                )
+
+                Spacer(modifier = Modifier.weight(1f))
+
+                Text(
+                    "${steps.count { it.state == State.COMPLETED }}/${steps.size}",
+                    style = MaterialTheme.typography.labelSmall
+                )
+
+                ArrowButton(modifier = Modifier.size(24.dp), expanded = expanded, onClick = null)
+            }
+        }
+
+        AnimatedVisibility(visible = expanded) {
+            Column(
+                modifier = Modifier.fillMaxWidth()
+            ) {
+                if (category != StepCategory.PATCHING) {
+                    steps.forEach { step ->
+                        val downloadProgress = step.downloadProgress?.collectAsStateWithLifecycle()
+
+                        SubStep(
+                            name = step.name,
+                            state = step.state,
+                            message = step.message,
+                            downloadProgress = downloadProgress?.value
+                        )
+                    }
+                } else {
+                    val currentPatch = remember(steps) {
+                        steps.find { it.state == State.RUNNING }?.name
+                    }
+                    val failedPatch = remember(steps) {
+                        steps.find { it.state == State.FAILED }?.name
+                    }
+
+                    val name = remember(steps) {
+                        when {
+                            steps.any { it.state == State.RUNNING } -> context.getString(R.string.applying_patch, currentPatch)
+                            steps.any { it.state == State.FAILED } -> context.getString(R.string.failed_to_apply_patch, failedPatch)
+                            steps.all { it.state == State.WAITING } -> context.getString(R.string.apply_patches)
+                            else -> context.resources.getQuantityString(R.plurals.applied_patches, steps.size, steps.size)
+                        }
+                    }
+
+                    val state = remember(steps) {
+                        when {
+                            steps.all { it.state == State.COMPLETED } -> State.COMPLETED
+                            steps.any { it.state == State.FAILED } -> State.FAILED
+                            steps.any { it.state == State.RUNNING } -> State.RUNNING
+                            else -> State.WAITING
+                        }
+                    }
+
+                    SubStep(
+                        name = name,
+                        state = state,
+                        message = steps.find { it.state == State.FAILED }?.message
+                    )
+                }
+            }
```

Comment:
Would it be possible to accomplish this in the worker by adding the ability to change the name of a step or is that just worse?

Reply:
Yes, but wouldn't it be possible to do this in the worker by continuously updating the name and state of a single step?

Reply:
How would we pass that to the UI?

Reply:
We could make `onProgress` accept a boolean to decide whether it should increment the index or not.
A separate callback could be used instead

Reply:
How would we get the name of the patch that is still applying tho?

Reply:
`Session.run` gets the patch object list, which contains the names. Wouldn't that work?